backend/stellar_withdrawal_validator.go

```diff
@@ -0,0 +1,119 @@
+		Title:  "Withdrawal Already Executed",
+		Status: http.StatusBadRequest,
+		Detail: "The withdrawal has already been executed.",
+	}
+	WithdrawalAssetInvalid = problem.P{
+		Type:   "withdrawal_asset_invalid",
+		Title:  "Withdrawal Asset Invalid",
+		Status: http.StatusBadRequest,
+		Detail: "Withdrawing the requested asset is not supported by the bridge." +
+			"Refund the deposit once the withdrawal period has expired.",
+	}
+	WithdrawalAmountInvalid = problem.P{
+		Type:   "withdrawal_amount_invalid",
+		Title:  "Withdrawal Amouont Invalid",
+		Status: http.StatusBadRequest,
+		Detail: "Withdrawing the requested amount is not supported by the bridge." +
+			"Refund the deposit once the withdrawal period has expired.",
+	}
+
+	ethereumTokenAddress = common.Address{}
+)
+
+// StellarWithdrawalValidator checks if it is possible to
+// withdraw a deposit to the ethereum bridge smart contract on
+// Stellar.
+type StellarWithdrawalValidator struct {
+	Store            *store.DB
+	WithdrawalWindow time.Duration
+}
+
+// StellarWithdrawalDetails includes metadata about the
+// validation result.
+type StellarWithdrawalDetails struct {
+	// Deadline is the deadline for executing the withdrawal
+	// transaction on Stellar.
+	Deadline time.Time
+	// LedgerSequence is the sequence number of the Stellar ledger
+	// for which the validation result is accurate.
+	LedgerSequence uint32
+}
+
+func (s StellarWithdrawalValidator) CanWithdraw(ctx context.Context, deposit store.EthereumDeposit) (StellarWithdrawalDetails, error) {
+	// TODO: add support for erc20 transfers
+	if !common.IsHexAddress(deposit.Token) ||
+		common.HexToAddress(deposit.Token) != ethereumTokenAddress {
+		return StellarWithdrawalDetails{}, WithdrawalAssetInvalid
+	}
+
+	// TODO: implement amount validation which is specific to the type of token
+	amount := &big.Int{}
+	_, ok := amount.SetString(deposit.Amount, 10)
+	if !ok || !amount.IsInt64() || amount.Cmp(big.NewInt(0)) <= 0 {
+		return StellarWithdrawalDetails{}, WithdrawalAmountInvalid
+	}
+
+	err := s.Store.Session.BeginTx(&sql.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  true,
+	})
+	if err != nil {
+		return StellarWithdrawalDetails{}, errors.Wrap(err, "error starting repeatable read transaction")
+	}
+	defer func() {
+		// explicitly ignore return value to make the linter happy
+		_ = s.Store.Session.Rollback()
+	}()
+
+	lastLedgerSequence, err := s.Store.GetLastLedgerSequence(context.Background())
+	if err != nil {
+		return StellarWithdrawalDetails{}, errors.Wrap(err, "error getting last ledger sequence")
+	}
+
+	lastLedgerCloseTime, err := s.Store.GetLastLedgerCloseTime(ctx)
+	if err != nil {
+		return StellarWithdrawalDetails{}, errors.Wrap(err, "error getting last ledger close time")
+	}
+	withdrawalDeadline := time.Unix(deposit.Timestamp, 0).Add(s.WithdrawalWindow)
+	if lastLedgerCloseTime.After(withdrawalDeadline) {
+		return StellarWithdrawalDetails{}, WithdrawalWindowExpired
+	}
```

Comment:
all though the refund for a Ethereum->Stellar transfer will occur on ethereum we will only authorize the ethereum refund transaction if the last ledger close time (on Stellar) is after the withdrawal deadline. 

the stellar withdrawal transaction can only occur if the stellar ledger is before the withdrawal deadline.

Because we are using the stellar ledger close time for both withdrawals and refunds it should be impossible for there to be a withdrawal and a refund.

Reply:
Can't Ethereum user refund without interaction with Starbridge validator or the validators' signatures are required? If the latter, then it makes sense.